plugins/openssl/openssl.cpp

```diff
@@ -0,0 +1,950 @@
+        EVP_CIPHER_CTX * decryptCtx = EVP_CIPHER_CTX_new();
+        if (!decryptCtx)
+            failOpenSSLError("EVP_CIPHER_CTX_new");
+
+        // Reserve buffers
+        __lenResult = len_ciphertext; // max size, may be changed later
+        MemoryBuffer resultBuffer(__lenResult);
+
+        try
+        {
+            int len = 0;
+            int plaintextLen = 0;
+
+            if (EVP_DecryptInit_ex(decryptCtx, cipher, nullptr, static_cast<byte *>(key.bufferBase()), static_cast<byte *>(iv.bufferBase())) != 1)
+                failOpenSSLError("EVP_DecryptInit_ex");
+
+            if (EVP_DecryptUpdate(decryptCtx, static_cast<byte *>(resultBuffer.bufferBase()), &len, static_cast<const byte *>(_ciphertext), len_ciphertext) != 1)
+                failOpenSSLError("EVP_DecryptUpdate");
+            plaintextLen = len;
+
+            if (EVP_DecryptFinal_ex(decryptCtx, static_cast<byte *>(resultBuffer.bufferBase()) + len, &len) != 1)
+                failOpenSSLError("EVP_DecryptFinal_ex");
+            plaintextLen += len;
+            __lenResult = plaintextLen;
+            __result = resultBuffer.detachOwn();
+        }
+        catch (...)
+        {
+            EVP_CIPHER_CTX_free(decryptCtx);
+            __lenResult = 0;
+            rtlFree(__result);
+            __result = nullptr;
+            throw;
+        }
+
+        EVP_CIPHER_CTX_free(decryptCtx);
+    }
+}
+
+// pk functions
+
+OPENSSL_API void OPENSSL_CALL pkRSASeal(ICodeContext *ctx, size32_t & __lenResult, void * & __result, size32_t len_plaintext, const void * _plaintext, bool isAll_pem_public_keys, size32_t len_pem_public_keys, const void * _pem_public_keys, const char * _algorithm_name)
+{
+    // Initial sanity check of our arguments
+    if (len_pem_public_keys == 0)
+        rtlFail(-1, "No public keys provided");
+
+    if (!isAll_pem_public_keys && len_plaintext > 0)
+    {
+        std::vector<EVP_PKEY *> publicKeys;
+        EVP_CIPHER_CTX * encryptCtx = nullptr;
+        byte ** encryptedKeys = nullptr;
+        MemoryBuffer iv;
+        MemoryBuffer ciphertext;
+
+        try
+        {
+            // Parse public keys and stuff them into a vector
+            const char * pubKeyPtr = static_cast<const char *>(_pem_public_keys);
+            const char * endPtr = pubKeyPtr + len_pem_public_keys;
+            while (pubKeyPtr < endPtr)
+            {
+                const size32_t keySize = *(reinterpret_cast<const size32_t *>(pubKeyPtr));
+                pubKeyPtr += sizeof(keySize);
+                publicKeys.push_back(pkeyCache.checkCache(keySize, pubKeyPtr, 0, nullptr));
+                pubKeyPtr += keySize;
+            }
+
+            // Load the cipher
+            const EVP_CIPHER * cipher = cipherCache.checkCache(_algorithm_name);
+
+            // Allocate memory for encrypted keys
+            size_t keyCount = publicKeys.size();
+            encryptedKeys = static_cast<byte **>(rtlMalloc(sizeof(byte *) * keyCount));
```

Comment:
minor: You could use
enryptedKeys = new byte * [keyCount];
and
delete [] encryptedKeys;

rtlMalloc and rtlFree need to be used for allocated data that is returned from the function.  The reason is to avoid conflicts between the debug and release versions of the c++ library.